File: backend/src/main/java/com/schoolfinder/api/Dtos.java
package com.schoolfinder.api;

import com.schoolfinder.domain.ReviewStatus;
import com.schoolfinder.domain.Role;
import com.schoolfinder.domain.SchoolCategory;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.List;
import org.springframework.data.domain.Page;

/** All API request/response DTOs as Java records. Import the nested ones directly, e.g.
 *  {@code import com.schoolfinder.api.Dtos.SchoolSummary;} */
public final class Dtos {

    private Dtos() {
    }

    // ---------- Schools ----------

    public record SchoolSummary(
        Long id,
        String name,
        SchoolCategory category,
        String city,
        String region,
        BigDecimal tuitionFee,
        String currency,
        String coverImageUrl,
        Double latitude,
        Double longitude,
        double averageRating,
        long ratingCount
    ) {
    }

    public record ProgramDto(
        Long id,
        String name,
        String faculty,
        String level,
        Integer durationMonths,
        BigDecimal tuitionFee
    ) {
    }

    public record ImageDto(Long id, String url, String caption) {
    }

    public record SchoolDetail(
        Long id,
        String name,
        SchoolCategory category,
        String description,
        String history,
        String city,
        String region,
        String address,
        Double latitude,
        Double longitude,
        BigDecimal tuitionFee,
        String currency,
        String website,
        String phone,
        String email,
        String coverImageUrl,
        double averageRating,
        long ratingCount,
        boolean favorite,
        List<ProgramDto> programs,
        List<ImageDto> images
    ) {
    }

    // ---------- Reviews ----------

    public record CreateReviewRequest(
        @Min(1) @Max(5) int rating,
        @Size(max = 4000) String comment
    ) {
    }

    public record ReviewDto(
        Long id,
        Long schoolId,
        String schoolName,
        String userDisplayName,
        int rating,
        String comment,
        ReviewStatus status,
        OffsetDateTime createdAt
    ) {
    }

    // ---------- Admin: schools ----------

    public record SchoolUpsertRequest(
        @NotBlank String name,
        SchoolCategory category,
        String description,
        String history,
        String city,
        String region,
        String address,
        Double latitude,
        Double longitude,
        BigDecimal tuitionFee,
        String currency,
        String website,
        String phone,
        String email,
        String coverImageUrl,
        Boolean approved
    ) {
        public SchoolUpsertRequest {
            if (currency == null) currency = "XAF";
            if (approved == null) approved = Boolean.TRUE;
        }
    }

    // ---------- Users / me ----------

    public record MeDto(Long id, String firebaseUid, String email, String displayName, Role role) {
    }

    public record UserDto(Long id, String email, String displayName, Role role, boolean active, OffsetDateTime createdAt) {
    }

    // ---------- Generic ----------

    public record PageResponse<T>(List<T> content, int page, int size, long totalElements, int totalPages) {
        public static <T> PageResponse<T> of(Page<T> p) {
            return new PageResponse<>(p.getContent(), p.getNumber(), p.getSize(), p.getTotalElements(), p.getTotalPages());
        }
    }

    public record CompareResponse(List<SchoolDetail> schools, Long cheapestSchoolId, Long highestRatedSchoolId) {
    }

    public record MetaResponse(
        String appName,
        String tagline,
        String country,
        String currency,
        double mapCenterLat,
        double mapCenterLng,
        List<String> categories,
        boolean firebaseEnabled
    ) {
    }

    public record ApiError(int status, String error, String message, OffsetDateTime timestamp) {
        public ApiError(int status, String error, String message) {
            this(status, error, message, OffsetDateTime.now());
        }
    }
}

File: backend/src/main/kotlin/com/schoolfinder/api/Dtos.kt
package com.schoolfinder.api

import com.schoolfinder.domain.ReviewStatus
import com.schoolfinder.domain.Role
import com.schoolfinder.domain.SchoolCategory
import jakarta.validation.constraints.Max
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Size
import java.math.BigDecimal
import java.time.OffsetDateTime

// ---------- Schools ----------

data class SchoolSummary(
    val id: Long,
    val name: String,
    val category: SchoolCategory,
    val city: String?,
    val region: String?,
    val tuitionFee: BigDecimal?,
    val currency: String,
    val coverImageUrl: String?,
    val latitude: Double?,
    val longitude: Double?,
    val averageRating: Double,
    val ratingCount: Long,
)

data class ProgramDto(
    val id: Long,
    val name: String,
    val faculty: String?,
    val level: String?,
    val durationMonths: Int?,
    val tuitionFee: BigDecimal?,
)

data class ImageDto(val id: Long, val url: String, val caption: String?)

data class SchoolDetail(
    val id: Long,
    val name: String,
    val category: SchoolCategory,
    val description: String?,
    val history: String?,
    val city: String?,
    val region: String?,
    val address: String?,
    val latitude: Double?,
    val longitude: Double?,
    val tuitionFee: BigDecimal?,
    val currency: String,
    val website: String?,
    val phone: String?,
    val email: String?,
    val coverImageUrl: String?,
    val averageRating: Double,
    val ratingCount: Long,
    val favorite: Boolean,
    val programs: List<ProgramDto>,
    val images: List<ImageDto>,
)

// ---------- Reviews ----------

data class CreateReviewRequest(
    @field:Min(1) @field:Max(5) val rating: Int,
    @field:Size(max = 4000) val comment: String? = null,
)

data class ReviewDto(
    val id: Long,
    val schoolId: Long,
    val schoolName: String?,
    val userDisplayName: String?,
    val rating: Int,
    val comment: String?,
    val status: ReviewStatus,
    val createdAt: OffsetDateTime,
)

// ---------- Admin: schools ----------

data class SchoolUpsertRequest(
    @field:NotBlank val name: String,
    val category: SchoolCategory,
    val description: String? = null,
    val history: String? = null,
    val city: String? = null,
    val region: String? = null,
    val address: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val tuitionFee: BigDecimal? = null,
    val currency: String = "XAF",
    val website: String? = null,
    val phone: String? = null,
    val email: String? = null,
    val coverImageUrl: String? = null,
    val approved: Boolean = true,
)

// ---------- Users / me ----------

data class MeDto(
    val id: Long,
    val firebaseUid: String,
    val email: String?,
    val displayName: String?,
    val role: Role,
)

data class UserDto(
    val id: Long,
    val email: String?,
    val displayName: String?,
    val role: Role,
    val active: Boolean,
    val createdAt: OffsetDateTime,
)

// ---------- Generic ----------

data class PageResponse<T>(
    val content: List<T>,
    val page: Int,
    val size: Int,
    val totalElements: Long,
    val totalPages: Int,
) {
    companion object {
        fun <T> of(p: org.springframework.data.domain.Page<T>) = PageResponse(
            content = p.content,
            page = p.number,
            size = p.size,
            totalElements = p.totalElements,
            totalPages = p.totalPages,
        )
    }
}

data class CompareResponse(
    val schools: List<SchoolDetail>,
    val cheapestSchoolId: Long?,
    val highestRatedSchoolId: Long?,
)

data class MetaResponse(
    val appName: String,
    val tagline: String,
    val country: String,
    val currency: String,
    val mapCenterLat: Double,
    val mapCenterLng: Double,
    val categories: List<String>,
    val firebaseEnabled: Boolean,
)
